### PR TITLE
[MERGE] calendar, google_calendar: Refactoring

### PR DESCRIPTION
**Pull request by LucasLefevre**
1/ Recurring Events
==============
TL;DR A bit of magic and ... Pouf ... all virtual events are now real records.

## Purpose
The current implementation of recurring events uses "virtual events" to represents all
events in the recurrence. Only one event is stored in the database. All other events are
dynamically built and sent to the web client which sees them as real records.
This implementation is old, difficult to read (6 years of changes and bug fixes) and
has become more and more difficult to maintain.
A fresh start would be welcome.

## Specification
When creating a recurring event, a `calendar.recurrence.rule` record should be created.
This model holds the rrule configuration and is responsible to manage events that result
from the rrule.
When a new recurrence is created, all resulting events are also created (stored in the
database). No more virtual events.
To avoid an explosion of events, a maximum of 720 events is created. With 720 events,
a daily recurrence lasts for 2 years and a weekly recurrence lasts for 15 years.
This strategy is used by Google Calendar and seems acceptable in most cases.
A cron job could eventually be introduced to generated more events as the end comes.

This allows to introduce an new rrule end type: 'Forever'. This is actually a shortcut
to create a recurrence running for 720 events.

Inspired by Google Calendar, new possibilities are introduced when modifying a recurring
event (e.g. change the name, add an attendee):
- Modify only this particular event (the event is still part of the recurrence)
- Modify this and following events (events are still in the same recurrence)
- Modify all events

In a similar way, when updating the rrule of an event (e.g. from every Monday to every
Tuesday), the user can choose to:
- Modify this and following events: the recurrence is split. The first part remains
  unchanged but now ends when the second recurrence begins. The second parts is the
  updated rrule.
- Modify only this particular event.
- Modify all events: Forbidden (same as Google Calendar)

Note: when events are "moved" because the rrule changed, the events are not actually
moved. They are unlinked and new events are created (See "Some design choices explained"
section).

### Where to store the rrule?

Two options were considered.
1) Store the rrule on an event, each event of the recurrence having a Many2one
to the parent_id, aka the Master Event of the recurrence.

2) Store the rrule on another model: `calendar.recurrence.rule`. Each event in the
recurrence having a Many2one to the recurrence record.

Both options have pros & cons. But there is no clear winner.
However the actual business logic of handling the recurrence creation/update
would be the same.
Where the rrule configuration is stored is the only main difference between both options.
And this is probably the easiest part of implementing recurring events.

With that in mind, option 2) is chosen. It allows to clearly separate recurrence
logic from the events themselves. It is also the "correct" way of modeling data to avoid
many empty columns for most records.

### Reusing events on rrule update
When an rrule is modified, events should also be updated.
In most cases, current events are unlinked and new events are created from scratch.
Only events exactly at the same time before and after the rrule update are kept.
Trying to reuse other events would require an obscure and arbitrary heuristic.
(Imagine an rrule every Monday that is changed to every Tuesday and every Friday.
What would you do?).
This implies that any change to a specific event (name change, attendee added, chatter
messages) is lost. This tradeoff seems acceptable as updating an rrule should not be
that frequent. Moreover, Google Calendar also works that way so why not Odoo?

### Reusing events on events shifts

On the calendar view, drag and drop an event to shift the entire recurrence.

One way to handle the recurrence shift is to apply the same timedelta to all events.
Now events are correctly positioned... except for events that were specifically
moved. Those outliers needs special handling.
This also introduces some behavior inconsistencies:
when an rrule is directly modified, events are not reused, but when the rrule is
modified by drag & dropping an event, events are reused. This option needs more code to
handle the shift and the behaviors are not consistent.

The chosen option is to find the rrule configuration from the dragged event (this is
easy) and update the recurrence with those new rrule values. This brings us back to
an rrule update (see above): code reusability yeah; one behavior to rule them all yeah.
One downside is that outliers are lost (but Google Calendar also works that way).

2/ Synchronization with Google Calendar
==============
This commit refactors calendar synchronization between Odoo and Google after the
main calendar application refactoring.

This refactoring takes advantage of two new features from the Google API:

- New way of synchronizing resources efficiently[1]
Incremental sync is performed repeatedly and updates Odoo with all the changes that
happened ever since the previous sync. Each time, Odoo provides the previous sync
token it obtained from Google and stores the new sync token from the response.

- Event metadata[2]
Ability to set hidden key-value pairs with an event, called extended properties.
These extended properties are used to store the related odoo event id and the Odoo
owner id (see known limitations)

### Known limitations
- Let A and B be two new users (no tokens available). A creates an event in Odoo and
invites B. A is the owner of the event (user_id). Now B authenticates to his Google Calendar
account and synchronizes his calendar. We cannot send the event to A's calendar since we
don't have any access to his Google Calendar. Hence the event his sent to B's calendar.
This leads to data de-synchronisation: The owner is A in Odoo but B in Google.
The "real" owner (user A) is stored in the Google event's metadata to be able to
reconcile the owner for following synchronizations.

- Let A and B be two users of Odoo and Google Calendar. And let the Google Calendar of B
be private (e.g. if A creates an event in Google Calendar and invites B, B won't see the
event in his calendar). If A creates an event in Odoo and invites B. The event is synced
to Google Calendar of A. Now B can see the event in Odoo but he can't see it in his
Google Calendar.

Task 2126717
PR #42031
PR Enterprise https://github.com/odoo/enterprise/pull/8006

[1] https://developers.google.com/calendar/v3/sync
[2] https://developers.google.com/calendar/extended-properties

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/calendar/models/calendar_event.py [hunk 1]
```diff
@@ -3,92 +3,46 @@
...
+    :returns: ocurrence
+
+    >>> get_weekday_occurence(date(2019, 12, 17))
+    3  # third Tuesday of the month
+
+    >>> get_weekday_occurence(date(2019, 12, 25))
+    -1  # last Friday of the month
+    """
+    occurence_in_month = math.ceil(date.day/7)
+    if occurence_in_month in {4, 5}:  # fourth or fifth week on the month -> last
```
Comment (tivisse) [inline addons/calendar/models/calendar_event.py]: Maybe explain this functional choice in the commit message too ;)

File: addons/calendar/models/calendar_event.py [hunk 2]
```diff
@@ -593,6 +449,7 @@ def create_attendees(self):
                     'event_id': meeting.id,
                 }
 
+                # TODO move this in the correct module
```
Comment (tivisse) [inline addons/calendar/models/calendar_event.py]: True ^^

File: addons/calendar/models/calendar_event.py [hunk 3]
```diff
@@ -978,133 +543,136 @@ def action_sendmail(self):
...
+        self.recurrency = True
+        to_update |= self.env['calendar.recurrence.rule'].create(recurrence_vals)
+        return to_update._apply_recurrence()
+
+    def _get_recurrence_params(self):
+        if not self:
+            return {}
+        event_date = self._get_start_date()
+        weekday_field_name = weekday_to_field(event_date.weekday())
+        get_weekday_occurence
```
Comment (tivisse) [inline addons/calendar/models/calendar_event.py]: Not very useful :-)

File: addons/calendar/models/calendar_recurrence_rule.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+def weekday_to_field(weekday_index):
+    return RRULE_WEEKDAY_TO_FIELD.get(weekday_index)
+
+
+class RecurrenceRule(models.Model):
+
+    _name = 'calendar.recurrence.rule'
+    _description = 'Event Recurrence Rule'
+
+    # TODO exdate?
```
Comment (tivisse) [inline addons/calendar/models/calendar_recurrence_rule.py]: What is it supposed to be ?

File: addons/web/static/src/js/views/list/list_model.js [hunk 5]
```diff
@@ -103,6 +103,7 @@ odoo.define('web.ListModel', function (require) {
             options = options || {};
             options.fetchRecordsWithGroups = true;
             return this._super(list, options).then(function (result) {
+                debugger
```
Comment (tivisse) [inline addons/web/static/src/js/views/list/list_model.js]: Blork

File: addons/calendar/tests/test_calendar.py [hunk 6]
```diff
@@ -162,13 +147,11 @@ def test_event_allday(self):
...
             'start': "2018-10-16 00:00:00",
             'start_date': "2018-10-16",
-            'start_datetime': False,
             'stop': "2018-10-18 00:00:00",
             'stop_date': "2018-10-18",
-            'stop_datetime': False,
             'allday': True,
         })
-
+        event.invalidate_cache()
```
Comment (LucasLefevre) [inline addons/calendar/tests/test_calendar.py]: Remove this.

File: addons/calendar/models/calendar_alarm.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            elif alarm.interval == "hours":
+                alarm.duration_minutes = alarm.duration * 60
+            elif alarm.interval == "days":
+                alarm.duration_minutes = alarm.duration * 60 * 24
+            else:
+                alarm.duration_minutes = 0
+
+    _interval_selection = {'minutes': 'Minutes', 'hours': 'Hours', 'days': 'Days'}
+
+    name = fields.Char('Name', translate=True, required=True)
```
Comment (tivisse) [inline addons/calendar/models/calendar_alarm.py]: Strange to translate something formatted like: ``%s - %s %s`` don't you think ?

File: addons/google_calendar/models/calendar.py [hunk 8]
```diff
@@ -1,48 +1,189 @@
...
+            'location': google_event.location,
+            'user_id': google_event.owner(self.env).id,
+            'privacy': google_event.visibility or self.default_get(['privacy'])['privacy'],
+            'attendee_ids': self._odoo_attendee_commands(google_event),
+            'alarm_ids': alarm_commands,
+            'recurrency': google_event.is_recurrent()
+        }
+
+        if google_event.reminders and google_event.reminders.get('useDefault'):
+            values
```
Comment (tivisse) [inline addons/google_calendar/models/calendar.py]: Such useful :p

File: addons/google_account/models/google_service.py [hunk 9]
```diff
@@ -164,9 +137,15 @@ def _do_request(self, uri, params={}, headers={}, type='POST', preuri="https://w
         ask_time = fields.Datetime.now()
         try:
             if type.upper() in ('GET', 'DELETE'):
-                res = requests.request(type.lower(), preuri + uri, params=params, timeout=TIMEOUT)
+                res = requests.request(type.lower(), preuri + uri, params=params, timeout=timeout)
             elif type.upper() in ('POST', 'PATCH', 'PUT'):
-                res = requests.request(type.lower(), preuri + uri, data=params, headers=headers, timeout=TIMEOUT)
+                print("__Request__")
```
Comment (tivisse) [inline addons/google_account/models/google_service.py]: Is it intended to dsplay all the request arguments on the logs ?

File: addons/google_account/models/google_service.py [hunk 10]
```diff
@@ -114,44 +114,17 @@ def _get_google_token_json(self, authorize_code, service):
...
-                    self.env(cur)['res.users'].browse(self.env.uid).sudo().write({'google_%s_rtoken' % service: False})
-            error_key = error.response.json().get("error", "nc")
-            _logger.exception("Bad google request : %s !", error_key)
-            error_msg = _("Something went wrong during your token generation. Maybe your Authorization Code is invalid or already expired [%s]") % error_key
-            raise self.env['res.config.settings'].get_config_warning(error_msg)
-
     # TODO JEM : remove preuri param, and rename type into method
     @api.model
-    def _do_request(self, uri, params={}, headers={}, type='POST', preuri="https://www.googleapis.com"):
+    def _do_request(self, uri, params={}, headers={}, type='POST', preuri="https://www.googleapis.com", timeout=TIMEOUT):
```
Comment (tivisse) [inline addons/google_account/models/google_service.py]: params={}  ----> NOOOOOOOOOO!

File: addons/calendar/models/calendar_event.py [hunk 11]
```diff
@@ -1162,175 +728,27 @@ def create(self, values):
...
-                    elif field == 'display_time':
-                        value = self._get_display_time(ls[1], ls[2], real.duration, real.allday)
-                    else:
-                        value = self.env.cache.get(real, f)
-                    self.env.cache.set(calendar, f, value)
-        return result
-
     @api.model
     def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=False, lazy=True):
         if 'date' in groupby:
```
Comment (tivisse) [inline addons/calendar/models/calendar_event.py]: The field date doesn't seem to exist anymore. I think we can get rid of the read_group override. ;)

File: addons/calendar/models/ir_attachment.py [hunk 12]
```diff
@@ -1,27 +0,0 @@
...
-
-from odoo.addons.calendar.models.calendar_event import get_real_ids
-
-
-class Attachment(models.Model):
-
-    _inherit = "ir.attachment"
-
-    @api.model
-    def _search(self, args, offset=0, limit=None, order=None, count=False, access_rights_uid=None):
```
Comment (tivisse) [inline addons/calendar/models/ir_attachment.py]: @LucasLefevre Qué brol quand même -_-

File: addons/calendar/models/calendar_attendee.py [hunk 13]
```diff
@@ -24,54 +28,72 @@ def _default_access_token(self):
...
+        return attendees
+
+    def unlink(self):
+        self._unsubscribe_partner()
+        return super().unlink()
+
+    def _subscribe_partner(self):
+        for event in self.event_id:
+            partners = (event.attendee_ids & self).partner_id - event.message_partner_ids
+            partners -= self.env.user.partner_id  # current user is automatically added as followers, don't add it twice.
```
Comment (tivisse) [inline addons/calendar/models/calendar_attendee.py]: I think that it's managed by the ``message_subscribe`` method, but it doesn't harm anybody. :)

File: addons/calendar/models/calendar_event.py [hunk 14]
```diff
@@ -695,8 +702,57 @@ def create(self, values):
...
+            public = [event for event in events if event.get('privacy') != 'private']
+            return private, public
+
+        def my_events(events):
+            """
+            :param events: list of event values (dict)
+            :return: tuple(my events, other events)
+            """
+            my = [event for event in events if event.get('user_id') and event.get('user_id')[0] == self.env.uid]
+            others = [event for event in events if not event.get('user_id') or event.get('user_id')[0] != self.env.uid]
```
Comment (tivisse) [inline addons/calendar/models/calendar_event.py]: Why not ``other = events - my`` ?

File: addons/calendar/tests/test_calendar.py (hunk 6 above)
Comment (tivisse) [reply addons/calendar/tests/test_calendar.py]: You sure ? The test is crashing without that line.

File: addons/calendar/models/calendar_event.py [hunk 15]
```diff
@@ -206,6 +205,7 @@ def _compute_is_highlighted(self):
     show_as = fields.Selection([('free', 'Free'), ('busy', 'Busy')], 'Show Time as', default='busy', required=True)
 
     # linked document
+    # LUL TODO use fields.Reference ?
```
Comment (tivisse) [inline addons/calendar/models/calendar_event.py]: Well it depends. Is an end user supposed to configure those field manually on the form view ? 

Or do we expect them to always use the calendar view + the action buttons,... that are offered to him?

If the second sentence is correct, I wouldn't use reference fields here. What do you think ?

File: addons/calendar/controllers/main.py [hunk 16]
```diff
@@ -3,30 +3,22 @@
 
 import werkzeug
 
-from odoo.api import Environment
 import odoo.http as http
 
 from odoo.http import request
-from odoo import SUPERUSER_ID
-from odoo import registry as registry_get
```
Comment (tivisse) [inline addons/calendar/controllers/main.py]: I would add the following explanation on the commit message (I let you rephrase it if you wish to keep it real).

Before we wanted to manage the multi-db thing.

Now we expect the users to work live on the same db. (That's the reason why in a multi-db environment, the database manager opens to force you choosing one). Everything is managed at the dispatcher side. It's not the role of the business code to think about the infrastructure. As for the actions buttons on the notification emails for example.

File: addons/google_account/controllers/main.py [hunk 17]
```diff
@@ -20,7 +20,9 @@ def oauth2callback(self, **kw):
 
         with registry(dbname).cursor() as cr:
             if kw.get('code'):
-                request.env(cr, request.session.uid)['google.%s' % service].set_all_tokens(kw['code'])
+                access_token, refresh_token, ttl = request.env['google.service']._get_google_tokens(kw['code'], service)
+                # LUL TODO only defined in google_calendar
```
Comment (tivisse) [inline addons/google_account/controllers/main.py]: ``_get_google_tokens`` is only called once. Maybe you could override the method on google_calendar right ?

File: addons/google_calendar/views/res_users_views.xml [hunk 18]
```diff
@@ -9,12 +9,16 @@
...
+                            <field name="google_calendar_rtoken" readonly="1"/>
+                            <field name="google_calendar_token" readonly="1"/>
+                            <field name="google_calendar_token_validity" readonly="1"/>
+                            <field name="google_calendar_sync_token" readonly="1"/>
+                            <field name="google_calendar_cal_id" readonly="1"/>
                         </group>
+                        <button
+                            string="Reset Account" class="btn btn-secondary"
+                            type="action" name="%(google_calendar_reset_account_action)d"
+                            context="{'default_user_id': active_id}"/>
```
Comment (tivisse) [inline addons/google_calendar/views/res_users_views.xml]: Maybe add the group_system on the button. Just in case.

File: addons/calendar/controllers/main.py (hunk 16 above)
Comment (LucasLefevre) [reply addons/calendar/controllers/main.py]: I'm not sure to understand all this multi-db thing

**Comment by tivisse** [comment]
@odony @antonylesuisse 

Here is a PR about the calendar refactoring (recurring events, google calendar synchro, ...).

If you wish to take a look a provide your feedback, help yourselves. 

The commit message is quite complete and explained the technical choices that were made + the known limitations of the current implementation.

Thanks.

**Comment by tivisse** [comment]
@robodoo r+

**Comment by tivisse** [comment]
@robodoo merge

File: addons/calendar/models/calendar_event.py [hunk 19]
```diff
@@ -297,18 +287,33 @@ def _compute_dates(self):
...
+    def _compute_stop(self):
+        # stop and duration fields both depends on the start field.
+        # But they also depends on each other.
+        # When start is updated, we want to update the stop datetime based on
+        # the *current* duration. In other words, we want: change start => keep the duration fixed and
+        # recompute stop accordingly.
+        # However, while computing stop, duration is marked to be recomputed. Calling `event.duration` would trigger
+        # its recomputation. To avoid this we manually mark the field as computed.
+        duration_field = self._fields['duration']
+        self.env.remove_to_compute(duration_field, self)
```
Comment (akazecki) [inline addons/calendar/models/calendar_event.py:310]: @LucasLefevre Hello, I know this is a very old commit, but I would appreciate a lot if you replied. I have a small question: is this line really necessary? It kind of breaks a customization I'm trying to do in the calendar, i.e. overriding `_compute_duration` to make it depend on other, custom fields (described in more detail in the issue I made: https://github.com/odoo/odoo/issues/273987).

What can I do to achieve this? I can see the code comments, but I noticed removing this line or adding `self.env.add_to_compute(duration_field, self)` at the end of the function works and doesn't seem to cause any errors, but I'm wondering if there's some other reason why this line is needed (and why an inverse function wasn't used here instead).

Comment (LucasLefevre) [reply addons/calendar/models/calendar_event.py:310]: Hello @akazecki 

TBH I don't remember. If it works without the line, then fine, remove it :wink: The ORM evolved since then, it might not be required anymore.

> and why an inverse function wasn't used here instead

An inverse doesn't update the values "live" in the form view. It might be the reason, but again, I don't remember. Feel free to check with the PO and change all of this.

EDIT: I thought you were from odoo R&D :sweat_smile: Then you can either work around the problem by manually mark the field as to be recomputed, or overwrite completely `_compute_duration` by not calling `super()`

Comment (akazecki) [reply addons/calendar/models/calendar_event.py:310]: @LucasLefevre Yep, that's probably what I'm going to do. Thank you for replying.
